### DISCUSSION TITLE
Get the correct x-forwarded-host when using a proxy

### DISCUSSION THREAD

**Discussion by cosbgn** [vercel/community#448] [Help]
Hi I have a monorepo with 2 projects.

- Project-A
- Project-B

The project A is the main project, while project B is simply a proxy deployment where I add custom domains for my clients. In that I have a vercel.json file with the following:

```
// project-b/vercel.json

{
	"rewrites": [ 
		{"source": "/", "destination": "https://project-a.com/api/go" },
		{"source": "/:id", "destination": "https://project-a.com/api/go" }
	]
}
```

So that any domain added to project-b will load the function `project-a/api/go.js`
This works perfectly, however on `project-a/api/go.js` I can't get the domain name.

`x-forwarded-host` is equal to `project-a.com` rather than the custom domain. 
Is there any way to do this?

**Comment by crowterli** [comment]
Hiya,

[x-forwarded-host](https://vercel.com/docs/concepts/edge-network/headers#x-forwarded-host) is equal to the host header (which I would expect to be the accessed host rather than another domain)

> This header represents the domain name as it was accessed by the client. If the deployment has been assigned to a preview URL or production domain and the client visits the domain URL, it contains the custom domain instead of the underlying deployment URL.

Are you able to elaborate on the use case here for using this header within your app?

**Reply by cosbgn** [reply]
Hi @crowterli, 

My use case is a URL shortener which splits the traffic between multiple pages (to create AB/Tests).
My app is structured like this: 
- Main app (example.com) - This is a static app with some /api routes. Since generated static content gets preference over rewrites I can't add a rewrite to this project to redirect traffic from proxy-example.com/123 to a specific /api endpoint so I have a second app for that.
- Second app: This is really an empty project with just a vercel.json file. The main domain is proxy-example.com and it has the following URL rewrite:

```
{"source": "/", "destination": "https://example.com/api/go" },
{"source": "/:id", "destination": "https://example.com/api/go" }
```
Pretty much is simply rewrites proxy-example.com/123 to example.com/api/go?id=123 - There I get the host and redirect accordingly.

My issue was that `x-forwarded-host` was always equal to example.com so I couldn't get the actual domain.
Surprisingly `x-forwarded-host` started returning the correct host by itself a few days ago. I also noticed that a new header `x-vercel-proxy-signature` appeared. However it doesn't seem to be documented anywhere.

So currently everything works. I just hope it keeps working this way. Should I be worried that something changes?

**Reply by crowterli** [reply]
Hi there.

I wouldn't expect it to change, those are pretty fundamental headers so they're unlikely to change. Do feel free to keep an eye on it though and if you need us to help further just let us know!

**Reply by harish-14** [reply]
Hi @crowterli,
Even I have a similar use case, but from yesterday we are noticing that `x-forwarded-host` points to vercel's deployment url instead of the domain what it is supposed to get forwarded to. I am dependent on `x-forwarded-host` and `referer` from headers, but it seems like sometimes `referer` too is missing from the headers. It was working fine till 20th Dec 2022, we are observing this behaviour from yesterday.

Any idea why it's happening and if there is any work around for this please let me know!

Thanks in advance

**Reply by evsasse** [reply]
@harish-14 I've seen the same problem around ~2022-12-21.

I've worked around it by updating my project to NextJS 13, and then using something like the following:
```typescript
export async function middleware(req: NextRequest) {
  // ...

  const requestHeaders = new Headers(req.headers);
  requestHeaders.set("x-workaround-forwarded-host", req.nextUrl.host);

  return NextResponse.rewrite(new URL("URL-OF-MY-OTHER-PROJECT"), {
    request: {
      headers: requestHeaders,
    },
  });

  // ...
}
```

Then on the other project, I've stopped reading the "x-forwarded-host" header, and started reading "x-workaround-forwarded-host" instead.

**Reply by harish-14** [reply]
thank you @evsasse, vercel itself fixed it for me as it was a bug from their end.....I would like to bookmark this and try this out....thank you for you kind response! cheers!

**Reply by MarcinCebula** [reply]
Any chance this fix could be related to my issue? My app stopped working yesterday, it stopped getting the subdomain when using zones. 

https://github.com/orgs/vercel/discussions/1317

**Reply by harish-14** [reply]
hi @MarcinCebula, we implemented @evsasse workaround and it worked for us....instead of relying on vercel headers we added our custom key in the middlware.....as the issue came up again on 2023-01-21
